File: back/DevC/src/test/java/com/back/devc/domain/interaction/report/entity/ReportTest.java
package com.back.devc.domain.interaction.report.entity;

import com.back.devc.domain.member.member.entity.Member;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

@DisplayName("Report")
class ReportTest {

    @Test
    @DisplayName("builder creates a pending report")
    void builder_createsPendingReport() {
        Member reporter = Member.createLocalMember("reporter@test.com", "password", "reporter");

        Report report = Report.builder()
                .reporter(reporter)
                .targetType(TargetType.POST)
                .targetId(10L)
                .reasonType("SPAM")
                .reasonDetail("Repeated promotion")
                .build();

        assertThat(report.reporter).isSameAs(reporter);
        assertThat(report.targetType).isEqualTo(TargetType.POST);
        assertThat(report.targetId).isEqualTo(10L);
        assertThat(report.reasonType).isEqualTo("SPAM");
        assertThat(report.reasonDetail).isEqualTo("Repeated promotion");
        assertThat(report.getStatus()).isEqualTo(ReportStatus.PENDING);
        assertThat(report.getProcessedByAdmin()).isNull();
        assertThat(report.getProcessedAt()).isNull();
    }

    @Test
    @DisplayName("processReport resolves the report and records admin")
    void processReport_resolvesReport() {
        Report report = pendingReport();
        Member admin = Member.createLocalAdminMember("admin@test.com", "password", "admin");

        report.processReport(admin);

        assertThat(report.getStatus()).isEqualTo(ReportStatus.RESOLVED);
        assertThat(report.getProcessedByAdmin()).isSameAs(admin);
        assertThat(report.getProcessedAt()).isNotNull();
    }

    @Test
    @DisplayName("rejectReport rejects the report and records admin")
    void rejectReport_rejectsReport() {
        Report report = pendingReport();
        Member admin = Member.createLocalAdminMember("admin@test.com", "password", "admin");

        report.rejectReport(admin);

        assertThat(report.getStatus()).isEqualTo(ReportStatus.REJECTED);
        assertThat(report.getProcessedByAdmin()).isSameAs(admin);
        assertThat(report.getProcessedAt()).isNotNull();
    }

    private Report pendingReport() {
        return Report.builder()
                .reporter(Member.createLocalMember("reporter@test.com", "password", "reporter"))
                .targetType(TargetType.COMMENT)
                .targetId(20L)
                .reasonType("ABUSE")
                .reasonDetail("Insulting content")
                .build();
    }
}

File: back/DevC/src/test/java/com/back/devc/domain/interaction/report/unit/ReportTest.kt
package com.back.devc.domain.interaction.report.unit

import com.back.devc.domain.interaction.report.entity.Report
import com.back.devc.domain.interaction.report.entity.Report.Companion.create
import com.back.devc.domain.interaction.report.entity.ReportStatus
import com.back.devc.domain.interaction.report.entity.TargetType
import com.back.devc.domain.member.member.entity.Member
import com.back.devc.domain.member.member.entity.Member.Companion.createLocalAdminMember
import com.back.devc.domain.member.member.entity.Member.Companion.createLocalMember
import org.assertj.core.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

@DisplayName("Report")
internal class ReportTest {
    @Test
    @DisplayName("create creates a pending report")
    fun create_createsPendingReport() {
        val reporter = createLocalMember(
            "reporter@test.com",
            "password",
            "reporter"
        )

        val report = create(
            reporter,
            TargetType.POST,
            10L,
            "SPAM",
            "Repeated promotion"
        )

        Assertions.assertThat<Member>(report.reporter).isSameAs(reporter)
        Assertions.assertThat<TargetType>(report.targetType).isEqualTo(TargetType.POST)
        Assertions.assertThat(report.targetId).isEqualTo(10L)
        Assertions.assertThat(report.reasonType).isEqualTo("SPAM")
        Assertions.assertThat(report.reasonDetail).isEqualTo("Repeated promotion")

        Assertions.assertThat<ReportStatus>(report.status).isEqualTo(ReportStatus.PENDING)
        Assertions.assertThat<Member?>(report.processedByAdmin).isNull()
        Assertions.assertThat(report.processedAt).isNull()
    }

    @Test
    @DisplayName("processReport resolves the report and records admin")
    fun processReport_resolvesReport() {
        val report = pendingReport()

        val admin = createLocalAdminMember(
            "admin@test.com",
            "password",
            "admin"
        )

        report.processReport(admin)

        Assertions.assertThat<ReportStatus>(report.status).isEqualTo(ReportStatus.RESOLVED)
        Assertions.assertThat<Member?>(report.processedByAdmin).isSameAs(admin)
        Assertions.assertThat(report.processedAt).isNotNull()
    }

    @Test
    @DisplayName("rejectReport rejects the report and records admin")
    fun rejectReport_rejectsReport() {
        val report = pendingReport()

        val admin = createLocalAdminMember(
            "admin@test.com",
            "password",
            "admin"
        )

        report.rejectReport(admin)

        Assertions.assertThat<ReportStatus>(report.status).isEqualTo(ReportStatus.REJECTED)
        Assertions.assertThat<Member?>(report.processedByAdmin).isSameAs(admin)
        Assertions.assertThat(report.processedAt).isNotNull()
    }

    private fun pendingReport(): Report {
        return create(
            createLocalMember(
                "reporter@test.com",
                "password",
                "reporter"
            ),
            TargetType.COMMENT,
            20L,
            "ABUSE",
            "Insulting content"
        )
    }
}